### PR TITLE
[feg][cwf][orc8r]Promote script to promote docker images from test repo to prod repo

### PR DESCRIPTION
**Pull request by ranj8work**
Signed-off-by: Ranjini Iyer <ranj8work@gmail.com>

## Summary
**First version of Promote script to promote docker images of feg, cwf, orc8r or docker from -test repo to -prod repo**
Added changes to push helm charts to the new repo as well.

## Test Plan
Testing is done by running this script on httpd docker image. Promoted its version 1.0 from feg-test to feg-prod. Screenshots are attached below: 

image
image

Testing for changes to push helm charts to the new repo:
image

## Additional Information

Changes made after the review comments from @tmdzk 
Screenshots of test evidence are captured below:
image

image

image

### CODE REVIEW

**Comment by ranj8work** [comment]
Updated changes to push helm charts to new jfrog as well:

Testing screenshot is pasted below:

image

File: orc8r/tools/helm/package.sh [hunk 1]
```diff
@@ -12,6 +12,7 @@
 # limitations under the License.
```
Comment (tmdzk) [inline orc8r/tools/helm/package.sh]: Let's separate the Helm changes from the more general promotion changes.

File: orc8r/tools/helm/package.sh [hunk 2]
```diff
@@ -12,6 +12,7 @@
...
 
 # package.sh packages and publishes orc8r helm charts to a private git repo
+# Updated to publish the helm charts to new jfrog artifactory repository as well
```
Comment (tmdzk) [inline orc8r/tools/helm/package.sh]: This script is meant to be use if you want to publish in your own helm registry. 
The logic should remain the same and we should still let people upload to their own registry (github repo)

The upload of the chart should be in a different script and in my opinion within CI, we have to stop to manual push packages/images/binaries

File: lte/gateway/release/promote.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/python3
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: Let's use fabric like we do in the current release process, to make it cleaner and fits our current tool suite

File: lte/gateway/release/promote.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            return keyfile.read().rstrip()
+    except IOError as e:
+        print(e)
+        abort("Please add the Jfrog API key for user %s to %s and re-run"
+              % (JFROG_USER, JFROG_APIKEY_PATH))
+
+JFROG_API_KEY = _get_jfrog_apikey()
+
+if __name__ == "__main__":
+    modules = [ 'docker', 'feg', 'cwf', 'orc8r', 'all' ]
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: From what I read If I pass `all` as modules it doesn't do what expected but just try to get all-test

File: lte/gateway/release/promote.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+       print(f"Invalid arguments. Usage: {sys.argv[0]} {modules} imageName version. Ex: promote.py feg gateway_go 1a4b2b30")
+       abort("Exiting...")
+
+    if sys.argv[1] not in modules:
+       print(f"Module {sys.argv[1]} not found in allowed list {modules}")
+       print(f"Usage: {sys.argv[0]} {modules} imageName version")
+       abort("Exiting...")
+
+    module = sys.argv[1]
+    imgName = sys.argv[2]
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: I don't think specifying the image name is needed. We need to promote all images that have the `imgVersion` available. 

like 
`promote.py feg 1a4b2b30`

File: lte/gateway/release/promote.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+              return -1
+    else:
+        print(f"INCORRECT API method {method} invoked. Please use GET or POST.")
+        return -1
+
+    return 0
+
+def validate(module):
+    print(f"Validating {imgName}:{imgVersion} in repop:{sourceRepo}")
+    status = callJFROGAPI(method='GET')
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: I think this function should be redesign. 
Calling callJfrogApi with a get shouldn't hardcode get packages. `callJfrogApi` should only be a function that call the api with 0 logic in it just the request.

File: lte/gateway/release/promote.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    return status
+
+
+def promote(module):
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: promote  with only the module argument should be able to promote all images of a module.

File: lte/gateway/release/promote.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    print(f"Promoting {imgName}:{imgVersion} from {sourceRepo} to {targetRepo}")
+    status = callJFROGAPI(method='POST', payload={ "targetRepo": targetRepo, "dockerRepository": imgName, "tag": imgVersion, "copy": "true" })
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: Same here callJfrogApi should only be doing the call, 0 logic there. 
payload could have been defined before for readability

File: lte/gateway/release/promote.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+import sys
+import requests
+from requests.auth import HTTPBasicAuth
+from fabric.utils import abort
+
+#################################
+# Global variables definitions
+#################################
+
+JFROG_URL = 'https://artifactory.magmacore.org/artifactory/api/docker'
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: As we will implement other promotions here the jfrog url shouldn't be tied to docker

File: lte/gateway/release/promote.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+targetRepo = ""
+imgName = "httpd"
+imgVersion = "1.0"
+jsonResponse = ""
+
+#######################################
+# User Defined functions definitions
+#######################################
+
+def callJFROGAPI(method='GET',
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: This function needs to be reworked to be more simple with 0 promotion logic. 
It should only be a function that calls the Jfrog Api

File: lte/gateway/release/promote.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+              print(f"HTTPError Occurred. Error=")
+              print(error)
+              return -1
+    elif method == 'POST':
+        try:
+           response = requests.post(url + '/' + sourceRepo + '/v2/promote', headers=reqHeaders, json=payload, auth=(JFROG_USER, JFROG_API_KEY))
+           if(response.status_code == 200):
+              print(f"SUCCESS promoting {imgName}:{imgVersion} from {sourceRepo} to {targetRepo}.")
+              print(f"Response Code={response.status_code}")
+              print(f"Response Body={response.content}")
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: you don't need to print that much information, just the first print is enough

File: lte/gateway/release/promote.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+              if imgVersion in jsonResponse["tags"]:
+                 print(f"{imgName}:{imgVersion} found in sourece repo:{sourceRepo}. Promoting it to {targetRepo}")
+              else:
+                 print(f"{imgName}:{imgVersion} NOT found in sourece repo:{sourceRepo}. NOT Promoting. Quitting...")
+                 return -1
+           else:
+              print(f"GET call failed with status Code={response.status_code}. Response={response.json()}")
+              return -1
+        except requests.exceptions.HTTPError as error:
+              print(f"HTTPError Occurred. Error=")
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: combine the 2 prints

File: lte/gateway/release/promote.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+           if(response.status_code == 200):
+              print(f"SUCCESS promoting {imgName}:{imgVersion} from {sourceRepo} to {targetRepo}.")
+              print(f"Response Code={response.status_code}")
+              print(f"Response Body={response.content}")
+           else:
+              print(f"Response Code={response.status_code}")
+              print(f"Response Body={response.json()}")
+              return -1
+        except requests.exceptions.HTTPError as error:
+              print("HTTPError Occurred. Error=")
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: combine the 2 prints

File: lte/gateway/release/promote.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+
+    status = validate(module)
+    if status == 0:
+       print(f"Validation of {imgName}:{imgVersion} in repop:{sourceRepo} is successful.")
+    else:
+       print(f"Validation of {imgName}:{imgVersion} in repop:{sourceRepo} FAILED. CANNOT Promote. Quitting...")
+       quit()
+
+    status = promote(module)
```
Comment (tmdzk) [inline lte/gateway/release/promote.py]: There is some logic missing. 
what if I want to promote all images? 
what if I only want the agw one? 

Let's do fab function:
1 for docker 
1 for agw 
1 for helm

**Review by tmdzk — CHANGES_REQUESTED** [review]
I think most of the code is there. 
But there are a few issues. 

Main feedback related to the promotion script:
1 -  let's use fabric in order to be able to split the differents function and keep using what magma devs and contributors are used to. 
2 - we should be able to promote all the images of a module 
3 - we should be able to promote all images of all modules at once

Please split the upload helm logic into a new script that we will integrate to CI

**Review by hcgatewood — APPROVED** [review]
Approving to unblock need for orc8r-approvers stamp. @tmdzk can handle full review of this PR

As an aside: agree with let's use fab

**Comment by hcgatewood** [comment]
Closing due to inactivity. Please reopen if desired @ranj8work

**Pull request closed without merging**